Use unique_id column to identify forms and modules

kaapstorm opened the pull request:
Context: [Asana](https://app.asana.com/0/1119457610578860/1113308365578031/f); PR #24151; [Doc](https://docs.google.com/document/d/1oWP_cJvcMZ7pO2aG6_Zru-4blpltCzI54C2tFhF_Nrc/edit#)

This approach differs from #24151 in that it leaves sheet names as they are, and uses a unique_id column, if is it available, to identify forms and modules. It also adds a unique_id column to single-sheet translations.

@emord @orangejenny @mkangia

kaapstorm committed:
Drop unused variables

kaapstorm committed:
Refactor BulkAppTranslationModulesAndFormsUpdater

kaapstorm committed:
Refactor BulkAppTranslationModuleUpdater

kaapstorm committed:
Refactor BulkAppTranslationFormUpdater

kaapstorm committed:
Add unique_id to single sheet

kaapstorm committed:
Use unique_id if we've got it

kaapstorm committed:
identifier is always a sheet name

kaapstorm committed:
Process modules-and-forms sheet first

kaapstorm committed:
Pull out _process_single_sheet

kaapstorm committed:
Map sheet names to unique IDs

kaapstorm committed:
Update tests

kaapstorm commented:
(Cancelling build because tests still need to be updated.)

orangejenny commented on corehq/apps/translations/app_translations/download.py:
```diff
@@ -23,33 +23,41 @@
             # Questions
             for row in get_form_question_rows([lang], form):
-                rows.append(prefix + row)
+                rows.append(prefix + row + [''])
```
This adds an empty string for unique id for form question rows, but the similar rows for modules (the loop over the `get_module_rows` up on new lines 40-42) don't get an empty string. Should they?

kaapstorm replied:
It's a little odd. It looks like form rows have the same number of columns as the headers, but module rows don't. Not sure why. I'll make them all the same.

kaapstorm replied:
Turns out module case list & case detail rows just didn't bother adding empty strings for media. I've made all the single-sheet rows the same length fwiw.

orangejenny commented on corehq/apps/translations/app_translations/utils.py:
```diff
@@ -161,11 +161,20 @@ def is_single_sheet(identifier):
+    elif row_type == 'Menu' or is_module_sheet(row_type):
+        return app.get_module_by_unique_id(unique_id)
+    else:
+        raise ValueError('Unknown row type "{}"'.format(row_type))
```
translate

kaapstorm replied:
Dropping `row_type` to use `sheet_name` instead. Also using consistent error messages regardless of whether we use `unique_id` or `sheet_name` to fetch the module/form.

orangejenny commented on corehq/apps/translations/app_translations/utils.py:
```diff
@@ -161,11 +161,20 @@ def is_single_sheet(identifier):
+        return get_module_from_sheet_name(app, sheet_name)
+
+
+def get_menu_or_form_by_unique_id(app, unique_id, row_type):
```
Why not always pass sheet name, since that's always available, instead of having row_type be either Module/Form or sheet name, which is a little overloaded?

(I don't feel strongly about this.)

kaapstorm replied:
Lol, just refactored a little, and I'm doing exactly that. I agree, it makes more sense.

orangejenny commented on corehq/apps/translations/app_translations/upload_app.py:
```diff
@@ -101,46 +101,45 @@ def process_bulk_app_translation_upload(app, workbook, expected_headers, lang=No
     return msgs
 
 
-def _process_rows(app, identifier, rows, sheet_name=None, lang=None):
```
Nice cleanup, this logic is definitely clearer now.

kaapstorm replied:
Good call. Some comments and a docstring would make this much more readable.

orangejenny commented on corehq/apps/translations/app_translations/upload_app.py:
```diff
@@ -86,20 +86,23 @@ def process_bulk_app_translation_upload(app, workbook, expected_headers, lang=No
 
         if is_single_sheet(sheet.worksheet.title):
             module_or_form = None
+            module_or_form_rows = {}
```
Curious what the reason for this change is. Ordering of error messages?

kaapstorm replied:
Not all rows have `unique_id` populated. I'm processing the rows that have a `unique_id` first, to build a dictionary that maps `sheet_name` to `unique_id`. That way I can look up subsequent forms and modules using their `unique_id` when the row only gives their `sheet_name`.

orangejenny replied:
aha, got it

orangejenny reviewed (approved):
Looks good, no blocking comments.

Thoughts on adding a test that exercises the new functionality? You might just be able to take an existing test and re-order some modules/forms before uploading test data.

kaapstorm committed:
Fix form unique_id in spreadsheets

kaapstorm committed:
Add column for "video by language"

kaapstorm committed:
Expand single-sheet columns for readability

kaapstorm committed:
Merge branch 'master' into nh/reach/use_existing_column

kaapstorm committed:
Stickler

orangejenny commented on corehq/apps/translations/app_translations/download.py:
```diff
@@ -27,37 +27,70 @@ def get_bulk_app_single_sheet_by_name(app, lang):
-                         [module.icon_by_language(lang)],
-                         [module.audio_by_language(lang)]) +
-            [module.unique_id]
+            prefix +
```
Much more readable :)

kaapstorm replied:
> this is making a longer method look even longer

:thinking: What about moving these out into separate functions? In fact, the "Name / menu media row" is identical for forms and modules, so reusing a function would be DRYer. That would make this method shorter, and the functions might be a bit more readable/self-documenting.

mkangia commented on corehq/apps/translations/app_translations/upload_app.py:
```diff
@@ -247,35 +258,22 @@ def update(self, rows):
+            if unique_id and sheet_name not in self.names_map:
+                self.names_map[sheet_name] = unique_id
+            elif not unique_id and sheet_name in self.names_map:
+                unique_id = self.names_map[sheet_name]
```
can you add a quick note on why we are checking on unique_id column here when we already have the names_map?

kaapstorm replied:
Yes, a comment would definitely make a difference here. (The "unique_id" column is not always populated. So line 265 builds `names_map` when we have a value for `unique_id`, so we can fetch it from `names_map` here when we don't have a value for it.)

mkangia replied:
its not clear why you would not have a unique_id available earlier. may be i just need to read better.

mkangia commented on corehq/apps/translations/app_translations/utils.py:
```diff
@@ -162,13 +205,14 @@ class BulkAppTranslationUpdater(object):
         '''
         super(BulkAppTranslationUpdater, self).__init__()
         self.app = app
+        self.names_map = names_map
```
how about sheet_names_to_id_map with a short documentation on the method def?

mkangia commented:
No blocking comments.

kaapstorm commented on corehq/apps/translations/app_translations/upload_app.py:
```diff
@@ -83,62 +86,70 @@ def process_bulk_app_translation_upload(app, workbook, expected_headers, lang=No
+    if is_form_sheet(sheet_name):
         try:
-            updater = BulkAppTranslationFormUpdater(app, identifier, lang=lang)
+            updater = BulkAppTranslationFormUpdater(app, sheet_name, names_map, lang=lang)
```
That's a good point! `BulkAppTranslationModulesAndFormsUpdater.update()` needs to be able [to populate `names_map`](https://github.com/dimagi/commcare-hq/blob/9d7214a59e9bd0555308290eec801c48fb3ab51e/corehq/apps/translations/app_translations/upload_app.py#L265), but `BulkAppTranslationModuleUpdater` and `BulkAppTranslationFormUpdater` don't -- they just need the ID.

mkangia replied:
So not changing it?

kaapstorm replied:
Changed

kaapstorm committed:
Break up function

kaapstorm committed:
Merge branch 'master' into nh/reach/use_existing_column

kaapstorm committed:
Use unique_id instead of names_map where possible

kaapstorm committed:
Comments and docstrings [skip ci]

kaapstorm commented on corehq/apps/translations/tests/test_bulk_app_translation.py:
```diff
@@ -821,6 +821,10 @@ def test_removing_form_translations(self):
             (MODULES_AND_FORMS_SHEET_NAME,
              (('Menu', 'menu1', 'Untitled Module', '', '', '', '', '', '765f110eb62fd26240a6d8bcdccca91b246b96c9'),
               ('Form', 'menu1_form1', 'Untitled Form', '', '', '', '', '', 'fffea2c32b7902a3efcb6b84c94e824820d11856'))),
+              # TODO: This unique ID ^^^ does not match the one in form_modifications/app.json. Why?
```
There are two things about these test failures that I don't understand. The first is mismatched form `unique_id` values. Try this:
1. Pick any app in your local environment. The smaller the better. Open its JSON source in your browser.
2. Open a Django shell, and get the app. e.g.:
```python
>>> app_id = '298b67da52ecef99b055077d710034ec'  # copied from the URL
>>> from corehq.apps.app_manager.dbaccessors import get_app
>>> app = get_app('demo', app_id)
>>> for m in app.get_modules():
...      print(m.unique_id)
...      for f in m.get_forms():
...          print(f.unique_id)

57d4ea1a2d09412a94123ace6de8d79c
0368dbd58e80484eb1d2acfc16d603de
b0e6333fd78e4578a1c1e48fd047a878
```
3. Search for the module's unique_id in your browser. It finds it.
4. Now search for a form's unique_id. Mine is not found. The form's unique_id values in the browser are different from the values in the shell. Why?!

kaapstorm commented on corehq/apps/translations/tests/test_bulk_app_translation.py:
```diff
@@ -821,6 +821,10 @@ def test_removing_form_translations(self):
+              # TODO: This unique ID ^^^ does not match the one in form_modifications/app.json. Why?
+              #
+              # TODO: This value comes from corehq/apps/translations/app_translations/download.py:80
+              #       But fetching the form the same way in corehq/apps/translations/app_translations/upload_app.py:316 fails.
```
The fact that the form's unique_id value in the shell is different from its value in the app definition might not matter, as long as `[f.unique_id for m in app.get_modules() for f in m.get_forms()]` is consistent. This is how the values of unique_id are set when the sheet is created in [download.py](https://github.com/dimagi/commcare-hq/blob/68907731201a0ed4e00fceb8b17d249f668ddac3/corehq/apps/translations/app_translations/download.py#L80), so it should find the forms in [upload_app.py](https://github.com/dimagi/commcare-hq/blob/4ad8430a699ae832da83ed1ef0c0144d48c10264/corehq/apps/translations/app_translations/upload_app.py#L316) when the sheet is uploaded. 

But this test is failing because in the test the value of `form.unique_id` *matches* the value in the app definition, i.e. the **app.json** file; it doesn't match the value in line 823 above ... even though the value in line 823 originally comes from `form.unique_id`.

What is happening?!

orangejenny commented:
Regarding form unique ids: the app json you look at in the browser is generated by [app_source](https://github.com/dimagi/commcare-hq/blob/218aec2f9b24aa8720cd6e0860381b4377282fa9/corehq/apps/app_manager/views/apps.py#L387) which calls `export_json` on the app which I imagine somewhere calls the module and form versions of `export_json`. [FormBase.export_json](https://github.com/dimagi/commcare-hq/blob/218aec2f9b24aa8720cd6e0860381b4377282fa9/corehq/apps/app_manager/models.py#L2374) and [ModuleDetailsMixin.export_json](https://github.com/dimagi/commcare-hq/blob/218aec2f9b24aa8720cd6e0860381b4377282fa9/corehq/apps/app_manager/models.py#L2370) both, by default, delete form unique IDs. That doesn't explain step by step what's going on (I don't know exactly what's going on) but I'd poke around that code for the answer.

orangejenny commented:
(edited code links in the comment above)

kaapstorm commented:
Thanks @orangejenny I'll check that out!

kaapstorm committed:
Drop undefined variable

kaapstorm committed:
4-space JSON

kaapstorm committed:
Set form unique_id

kaapstorm committed:
Add test for moved modules

It is hard to tell from app.json, but the Registration module and the
Stethoscope module have been swapped around. This test shows that "What
likeness dost this have?" is successfully translated to "What does this
look like?"

kaapstorm commented:
Fixed test failures. (Then fixed some more test failures.) Then, finally, added a test for moved modules.

orangejenny commented on corehq/apps/translations/tests/test_bulk_app_translation.py:
```diff
@@ -37,6 +37,89 @@
 from corehq.util.workbook_json.excel import get_workbook, WorkbookJSONReader
 
 
+EXCEL_HEADERS = (
```
Now that this is file-level data, it would be helpful to either rename these variables or add a comment mentioning which tests they're relevant to. This file is a mess of different test apps (I have greatly contributed to it...).

kaapstorm replied:
I think this is a great idea. I've merged so that GS can start testing. (I see I've *just* missed today's deploy.) I'll follow up in a separate PR.

orangejenny commented on corehq/apps/translations/tests/test_bulk_app_translation.py:
```diff
@@ -805,6 +888,20 @@ def test_unchanged_upload(self):
+
+    The upload is an unchanged download.
+    """
+    file_path = "data", "bulk_app_translation", "moved_module"
```
Definitely feel free to call this out of scope, but given that testing a moved module *could* test a pretty basic app, what would you think of using `AppFactory` instead of an external `app.json`? The JSON-based tests are harder to debug and update.

kaapstorm replied:
Another great idea!

mkangia commented:
Thanks for getting this through @kaapstorm ..

kaapstorm commented:
@dimagi/product This change allows app translations to be uploaded successfully after forms and modules in that app have been rearranged.
